File: mcp/src/tools/wiki.ts
/**
 * tools/wiki.ts
 * evor_wiki_add          — persist LessonEntry to run wiki + cross-run index.
 *                          Dedup: exact lesson_id OR near-dup (same node +
 *                          approach_family + obs[:100]) replaces in-place.
 *                          Cap: corpus capped at EVOR_WIKI_MAX_ENTRIES (default 500)
 *                          most-recent entries — atomic write via tmp→rename.
 * evor_wiki_get_relevant — TF-IDF semantic slice with lazy corpus cache keyed
 *                          by (indexPath, mtime); invalidates on every write.
 * evor_wiki_query        — legacy keyword search (prefer get_relevant).
 *
 * Implemented in pure TypeScript to avoid a Python subprocess dependency
 * for simple file I/O + keyword scoring.  Mirrors CompoundingWiki in
 * harness/evor/wiki.py exactly (same layout, same ranking logic).
 */

import {
  existsSync,
  mkdirSync,
  readFileSync,
  renameSync,
  statSync,
  writeFileSync,
} from "fs";
import { join } from "path";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { LessonEntry, LessonEntrySchema, ApproachFamilySchema } from "../contracts.js";
import { resolveRunPaths, getEvorRoot } from "../run-store.js";
import { ok } from "../tool-result.js";

// ── Markdown renderer (mirrors wiki.py._render_lesson) ────────────────────

function renderLesson(entry: LessonEntry): string {
  const lines: string[] = [
    `# Lesson: ${entry.lesson_id}`,
    "",
    `**Node:** ${entry.node_id}  |  **Run:** ${entry.run_id}  |  **Mission:** ${entry.mission_id}`,
    `**Family:** ${entry.approach_family}  |  **Verdict:** ${entry.hypothesis_verdict}`,
    `**Created:** ${entry.created_at}`,
    "",
    "## Observation",
    "",
    entry.observation,
    "",
  ];

  if (entry.root_cause) {
    lines.push("## Root Cause", "", entry.root_cause, "");
  }

  lines.push("## Actionable Lesson", "", entry.actionable_lesson, "");

  if (entry.telemetry_evidence) {
    lines.push("## Telemetry Evidence", "", entry.telemetry_evidence, "");
  }

  if (entry.citations.length > 0) {
    lines.push("## Citations", "");
    for (const cit of entry.citations) {
      lines.push(`- ${cit}`);
    }
    lines.push("");
  }

  if (entry.tags.length > 0) {
    lines.push(`**Tags:** ${entry.tags.join(", ")}`, "");
  }

  return lines.join("\n");
}

// ── Dedup / prune helpers ─────────────────────────────────────────────────

/** Max cross-run index entries. Override via EVOR_WIKI_MAX_ENTRIES env. */
function getWikiMaxEntries(): number {
  return parseInt(process.env.EVOR_WIKI_MAX_ENTRIES ?? "500", 10);
}

/**
 * Near-duplicate fingerprint.
 * Two entries are near-duplicates when the same experiment node emits
 * another lesson about the same approach family with the same observation
 * prefix (first 100 chars, lowercased) — a loop-artifact dedup guard.
 */
function nearDupKey(e: LessonEntry): string {
  return `${e.node_id}\0${e.approach_family}\0${e.observation.slice(0, 100).toLowerCase()}`;
}

/**
 * Read index.jsonl, apply dedup, cap, and write back atomically (tmp→rename).
 *
 * Dedup policy (first match wins — exact beats near-dup):
 *   - Exact:    same lesson_id     → replace in-place (idempotent re-run)
 *   - Near-dup: same node+family+obs[:100] → replace in-place (loop artifact)
 *
 * Cap policy: keep WIKI_MAX_ENTRIES most-recent entries by created_at (ISO sort).
 * Dropped entries lose their spot permanently; the MD files are left on disk.
 */
function pruneAndWrite(indexPath: string, newEntry: LessonEntry): void {
  const entries: LessonEntry[] = [];

  if (existsSync(indexPath)) {
    for (const line of readFileSync(indexPath, "utf8").split("\n")) {
      const trimmed = line.trim();
      if (!trimmed) continue;
      try {
        entries.push(LessonEntrySchema.parse(JSON.parse(trimmed)));
      } catch {
        continue; // skip malformed lines
      }
    }
  }

  const ndKey = nearDupKey(newEntry);
  let replaced = false;
  for (let i = 0; i < entries.length; i++) {
    if (entries[i].lesson_id === newEntry.lesson_id || nearDupKey(entries[i]) === ndKey) {
      entries[i] = newEntry;
      replaced = true;
      break;
    }
  }
  if (!replaced) {
    entries.push(newEntry);
  }

  // Cap: keep most-recent N
  const max = getWikiMaxEntries();
  if (entries.length > max) {
    entries.sort((a, b) => b.created_at.localeCompare(a.created_at));
    entries.splice(max);
  }

  // Atomic write: write to .tmp then rename (POSIX atomic)
  const tmp = `${indexPath}.tmp.${process.pid}`;
  writeFileSync(tmp, entries.map((e) => JSON.stringify(e)).join("\n") + "\n", "utf8");
  renameSync(tmp, indexPath);
  // Invalidate explicitly. mtime cannot be relied on: two writes inside the same
  // millisecond share a timestamp, and the reader would serve the pre-write corpus.
  bumpWikiGeneration();
}

// ── TF-IDF helpers ────────────────────────────────────────────────────────

/** Tokenize text: split on non-alpha chars, lowercase, drop tokens shorter than 2 chars. */
function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .split(/[^a-z]+/)
    .filter((t) => t.length >= 2);
}

/** Build a term-frequency map from a token list. */
function termFreq(tokens: string[]): Map<string, number> {
  const tf = new Map<string, number>();
  for (const t of tokens) {
    tf.set(t, (tf.get(t) ?? 0) + 1);
  }
  return tf;
}

/** Cosine similarity between two TF-IDF vectors (both represented as Map<term, weight>). */
function cosineSim(a: Map<string, number>, b: Map<string, number>): number {
  let dot = 0;
  let normA = 0;
  let normB = 0;
  for (const [term, wa] of a) {
    const wb = b.get(term) ?? 0;
    dot += wa * wb;
    normA += wa * wa;
  }
  for (const wb of b.values()) {
    normB += wb * wb;
  }
  if (normA === 0 || normB === 0) return 0;
  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

// ── IDF corpus cache ──────────────────────────────────────────────────────

/**
 * Cached corpus for one index file.
 *
 * entryVecs: per-entry TF-IDF vectors pre-computed from the corpus IDF map.
 * These are IDF-dependent and do not change while the file is unchanged.
 * Query vectors are built per-call (cheap: just the query tokens × IDF lookup).
 *
 * Cache key: (indexPath → mtime + generation).
 *
 * mtime ALONE IS NOT SUFFICIENT, which the original design assumed: "a write via
 * pruneAndWrite (tmp→rename) always produces a fresh mtime". It does not. mtimeMs
 * is millisecond-resolution and the kernel's timestamp source is coarser still —
 * measured on this host, two immediate writes to the same file produced identical
 * `mtime_ns`. So a wikiAdd followed by a wikiGetRelevant inside the same tick
 * served the PREVIOUS corpus, silently missing the lesson just written.
 *
 * That is a data-loss path, not a performance detail: Sage writes a finding and
 * Mutagen queries the wiki in the same tick, so the reader can miss exactly the
 * lesson the writer just produced — with no error anywhere.
 *
 * The generation counter is bumped by every in-process write, so invalidation no
 * longer depends on clock resolution. mtime is retained as the secondary key so
 * writes from ANOTHER process still invalidate.
 */
interface WikiCache {
  mtime: number;
  generation: number;
  entries: LessonEntry[];
  entryVecs: Map<string, number>[];
  df: Map<string, number>;
  N: number;
}

const idfCache = new Map<string, WikiCache>();

/** Bumped on every in-process write; part of the cache key. See WikiCache. */
let wikiGeneration = 0;

/** Invalidate the corpus cache for a path after a write. */
function bumpWikiGeneration(): void {
  wikiGeneration++;
}

/** Hit/miss counters exposed for tests. */
export const _wikiCacheStats = { hits: 0, misses: 0 };

/**
 * Clear the in-process IDF cache and reset counters.
 * Call in test beforeEach to ensure each test starts from a clean state.
 */
export function _resetWikiCache(): void {
  idfCache.clear();
  _wikiCacheStats.hits = 0;
  _wikiCacheStats.misses = 0;
  wikiGeneration = 0;
}

/**
 * Load (or cache-hit) the parsed corpus, document-frequency map, and
 * pre-computed per-entry TF-IDF vectors for indexPath.
 *
 * Returns an empty cache object if the file does not exist.
 */
function loadCorpus(indexPath: string): WikiCache {
  let mtime = 0;
  try {
    mtime = statSync(indexPath).mtimeMs;
  } catch {
    return { mtime: 0, generation: wikiGeneration, entries: [], entryVecs: [], df: new Map(), N: 0 };
  }

  const cached = idfCache.get(indexPath);
  if (cached && cached.mtime === mtime && cached.generation === wikiGeneration) {
    _wikiCacheStats.hits++;
    return cached;
  }
  _wikiCacheStats.misses++;

  // Cold load
  const entries: LessonEntry[] = [];
  for (const line of readFileSync(indexPath, "utf8").split("\n")) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    try {
      entries.push(LessonEntrySchema.parse(JSON.parse(trimmed)));
    } catch {
      continue;
    }
  }

  const entryTokensList: string[][] = entries.map((e) =>
    tokenize(
      [e.observation, e.actionable_lesson, e.root_cause ?? "", e.tags.join(" ")].join(" ")
    )
  );

  // Document frequency
  const df = new Map<string, number>();
  for (const tokens of entryTokensList) {
    for (const term of new Set(tokens)) {
      df.set(term, (df.get(term) ?? 0) + 1);
    }
  }

  const N = entries.length;

  // Pre-compute per-entry TF-IDF vectors using the corpus IDF
  const entryVecs: Map<string, number>[] = entryTokensList.map((tokens) => {
    const tf = termFreq(tokens);
    const vec = new Map<string, number>();
    for (const [term, count] of tf) {
      const idf = Math.log((N + 1) / ((df.get(term) ?? 0) + 1)) + 1;
      vec.set(term, count * idf);
    }
    return vec;
  });

  const result: WikiCache = { mtime, generation: wikiGeneration, entries, entryVecs, df, N };
  idfCache.set(indexPath, result);
  return result;
}

// ── Core logic (exported for tests) ──────────────────────────────────────

/**
 * Persist a LessonEntry to:
 *   - `run_dir/wiki/<lesson_id>.md`      (per-run copy)
 *   - `evor_root/wiki/<lesson_id>.md`    (cross-run copy)
 *   - `evor_root/wiki/index.jsonl`       (cross-run searchable index — bounded)
 *
 * Dedup: same lesson_id or same (node_id + approach_family + obs[:100])
 * replaces the existing entry rather than appending.
 * Cap: at most EVOR_WIKI_MAX_ENTRIES (default 500) entries kept.
 */
/** What a citation resolver reports back about one identifier (item 5.4). */
export type CitationResolution = {
  resolved: boolean;
  title?: string;
  /** False when the resolved work does not support the claim it is cited for. */
  supports_claim?: boolean;
};

export function wikiAdd(
  runId: string,
  entry: LessonEntry,
  missionId?: string,
  opts?: { resolveCitation?: (citation: string) => CitationResolution }
): { lessonId: string; indexPath: string; citation_status?: Record<string, CitationResolution> } {
  const paths = resolveRunPaths(runId, missionId);
  const evorRoot = getEvorRoot();
  const wikiRoot = join(evorRoot, "wiki");

  // ── Item 5.4 (N-01/N-04): resolve citations BEFORE persisting ───────────
  //
  // The only signal of citation identity in the whole pipeline was the junior's
  // own `urls_verified: true`, asserted about its own work, checked by nothing.
  // Three of twenty sampled citations were misattributed: CBAM was credited to
  // arXiv 2006.05595, which is "Fitted Q-Learning for Relational Domains".
  //
  // The resolver is INJECTED rather than called directly, so this path is
  // testable without a network and so the server — not the agent — owns the
  // verdict. No resolver supplied means no verification was performed, and the
  // result says so rather than defaulting to verified: absence of a failure
  // verdict is not evidence of integrity.
  let citation_status: Record<string, CitationResolution> | undefined;
  if (opts?.resolveCitation) {
    citation_status = {};
    for (const citation of entry.citations ?? []) {
      try {
        citation_status[citation] = opts.resolveCitation(citation);
      } catch (err) {
        citation_status[citation] = { resolved: false, title: `resolver error: ${String(err)}` };
      }
    }
  }

  const rendered = renderLesson(entry);

  // Cross-run wiki
  mkdirSync(wikiRoot, { recursive: true });
  writeFileSync(join(wikiRoot, `${entry.lesson_id}.md`), rendered, "utf8");
  const indexPath = join(wikiRoot, "index.jsonl");
  pruneAndWrite(indexPath, entry);

  // Per-run wiki
  const runWikiDir = join(paths.runDir, "wiki");
  mkdirSync(runWikiDir, { recursive: true });
  writeFileSync(join(runWikiDir, `${entry.lesson_id}.md`), rendered, "utf8");

  return { lessonId: entry.lesson_id, indexPath , citation_status };
}

/**
 * Keyword search over `evor_root/wiki/index.jsonl`.
 *
 * Scoring (mirrors wiki.py.query):
 *   - Each keyword hit in observation + actionable_lesson + tags + root_cause
 *     counts once per occurrence.
 *   - Lessons with zero hits are excluded (unless query is empty).
 *   - Sort: most hits first, then newest created_at desc.
 *
 * Filters: `family` and `confirmedOnly` applied before scoring.
 *
 * Search is one tool: evor_wiki_query with mode=semantic|keyword.
 */
export function wikiQuery(
  query: string,
  opts?: {
    family?: z.infer<typeof ApproachFamilySchema>;
    confirmedOnly?: boolean;
    limit?: number;
  }
): LessonEntry[] {
  const evorRoot = getEvorRoot();
  const indexPath = join(evorRoot, "wiki", "index.jsonl");

  if (!existsSync(indexPath)) {
    return [];
  }

  const keywords = query.split(/\s+/).filter(Boolean).map((k) => k.toLowerCase());
  const limit = opts?.limit ?? 10;

  const scored: Array<{ hits: number; createdAt: string; entry: LessonEntry }> = [];

  for (const line of readFileSync(indexPath, "utf8").split("\n")) {
    const trimmed = line.trim();
    if (!trimmed) continue;

    let entry: LessonEntry;
    try {
      entry = LessonEntrySchema.parse(JSON.parse(trimmed));
    } catch {
      continue;
    }

    // Filters
    if (opts?.family && entry.approach_family !== opts.family) continue;
    if (opts?.confirmedOnly && entry.hypothesis_verdict !== "confirmed") continue;
    // Item 5.3: a superseded entry is not a confirmed lesson, whatever its
    // verdict field still says. The r1 latency claim was falsified twice and
    // then cited 23 times across the tree, because nothing on the retrieval path
    // could tell a refuted entry from a standing one.
    if (opts?.confirmedOnly && (entry as { superseded_by?: string }).superseded_by) continue;

    // Keyword scoring
    if (keywords.length === 0) {
      scored.push({ hits: 0, createdAt: entry.created_at, entry });
      continue;
    }

    const blob = [
      entry.observation,
      entry.actionable_lesson,
      entry.root_cause ?? "",
      entry.tags.join(" "),
    ]
      .join(" ")
      .toLowerCase();

    const hits = keywords.reduce((sum, kw) => {
      let count = 0;
      let pos = 0;
      while ((pos = blob.indexOf(kw, pos)) !== -1) {
        count++;
        pos += kw.length;
      }
      return sum + count;
    }, 0);

    if (hits > 0) {
      scored.push({ hits, createdAt: entry.created_at, entry });
    }
  }

  // Sort: most hits first, then newest first
  scored.sort((a, b) => {
    if (b.hits !== a.hits) return b.hits - a.hits;
    return b.createdAt.localeCompare(a.createdAt);
  });

  return scored.slice(0, limit).map((s) => s.entry);
}

/**
 * Retrieve the k most semantically-relevant wiki lessons for a given context string.
 *
 * Uses smooth TF-IDF (sklearn default: IDF = log((N+1)/(df+1)) + 1) and cosine
 * similarity to rank entries. Per-entry TF-IDF vectors are cached against the
 * index file's mtime — a write via wikiAdd always busts the cache.
 *
 * Entries with zero cosine similarity are excluded.
 * Sort order: highest similarity first; newest created_at as tiebreaker.
 */
export function wikiGetRelevant(context: string, k: number): LessonEntry[] {
  const evorRoot = getEvorRoot();
  const indexPath = join(evorRoot, "wiki", "index.jsonl");

  const { entries, entryVecs, df, N } = loadCorpus(indexPath);
  if (entries.length === 0) return [];

  // Build query TF-IDF vector using cached IDF values
  const queryTokens = tokenize(context);
  if (queryTokens.length === 0) return [];

  const queryTf = termFreq(queryTokens);
  const queryVec = new Map<string, number>();
  for (const [term, count] of queryTf) {
    const idf = Math.log((N + 1) / ((df.get(term) ?? 0) + 1)) + 1;
    queryVec.set(term, count * idf);
  }

  if (queryVec.size === 0) return [];

  // Score each entry using pre-computed entry vectors
  const scored: Array<{ sim: number; createdAt: string; entry: LessonEntry }> = [];
  for (let i = 0; i < entries.length; i++) {
    const sim = cosineSim(queryVec, entryVecs[i]);
    if (sim > 0) {
      scored.push({ sim, createdAt: entries[i].created_at, entry: entries[i] });
    }
  }

  // Sort: highest similarity first, then newest created_at as tiebreaker
  scored.sort((a, b) => {
    if (b.sim !== a.sim) return b.sim - a.sim;
    return b.createdAt.localeCompare(a.createdAt);
  });

  return scored.slice(0, k).map((s) => s.entry);
}

// ── Tool registrations ────────────────────────────────────────────────────

export function registerWikiTools(server: McpServer): void {
  // ── evor_wiki_add ─────────────────────────────────────────────────────────
  server.tool(
    "evor_wiki_add",
    [
      "Write a LessonEntry to wiki/<lesson_id>.md and update .evor/wiki/index.jsonl.",
      "Dedup: same lesson_id OR same (node_id + approach_family + obs[:100]) replaces",
      "the existing entry instead of appending (prevents loop-duplicates).",
      "Cap: corpus is bounded at EVOR_WIKI_MAX_ENTRIES (default 500) most-recent entries.",
    ].join(" "),
    {
      run_id: z.string().describe("Active run identifier"),
      entry: LessonEntrySchema.describe("LessonEntry to persist"),
    },
    async ({ run_id, entry }) => {
      const missionId = process.env.EVOR_MISSION_ID;
      const { lessonId } = wikiAdd(run_id, entry, missionId);
      return {
        content: [
          {
            type: "text" as const,
            text: JSON.stringify({ ok: true, lesson_id: lessonId, run_id }),
          },
        ],
      };
    }
  );

  // ── evor_wiki_query ───────────────────────────────────────────────────────
  server.tool(
    "evor_wiki_query",
    [
      "Search wiki lessons across runs.",
      "mode=semantic ranks by TF-IDF relevance to a context string (approach, metric,",
      "error, architecture) and surfaces related concepts without exact keyword match;",
      "mode=keyword filters by exact phrase, approach_family and verdict.",
    ].join(" "),
    {
      // One tool, one enum, replacing two whose descriptions each told the caller
      // to use the other. Cross-recommendation in prose is a design smell: the
      // choice belongs in the interface, where it cannot drift from the behaviour
      // and costs no description bytes to restate (rubric rule 1).
      //
      // evor_wiki_get_relevant was folded in rather than the reverse because it
      // had ZERO references in agents/ or skills/, while evor_wiki_query is
      // referenced in six files — removing the referenced one would have broken
      // callers for a cosmetic gain.
      mode: z
        .enum(["semantic", "keyword"])
        .default("semantic")
        .describe("semantic = TF-IDF relevance ranking; keyword = exact filtering"),
      query: z.string().describe("Context string (semantic) or keyword/phrase (keyword)"),
      approach_family: ApproachFamilySchema.optional().describe("keyword mode: filter to an approach family"),
      confirmed_only: z.boolean().optional().describe("keyword mode: only hypothesis_verdict=confirmed"),
      limit: z.number().int().positive().optional().describe("Max results (semantic default 5, keyword default 10)"),
      // Accepted and ignored: never reached the query functions. Kept optional so
      // a caller still passing it is not rejected (PM3).
      run_id: z.string().optional().describe("Unused; accepted for backward compatibility"),
    },
    async ({ mode, run_id, query, approach_family, confirmed_only, limit }) => {
      const lessons =
        mode === "keyword"
          ? wikiQuery(query, { family: approach_family, confirmedOnly: confirmed_only, limit })
          : wikiGetRelevant(query, limit ?? 5);
      return ok({ run_id, mode, query, count: lessons.length, lessons });
    }
  );
}

File: harness/evor/wiki.py
"""
CompoundingWiki — per-run and cross-run lesson store (Addendum v2).

Layout on disk:
  .evor/wiki/
    index.jsonl              — append-only LessonEntry index (cross-run)
    <lesson-id>.md           — rendered markdown for each lesson (cross-run)
  .evor/runs/<mission>/<run-id>/wiki/
    <lesson-id>.md           — per-run copy (for run-scoped browsing)

add()   writes to both locations.
query() scans .evor/wiki/index.jsonl; filter by family; rank by recency.
load_context() returns top-N lessons by keyword relevance for mission startup.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Optional

from evor.contracts import ApproachFamily, LessonEntry


class CompoundingWiki:
    """Append-only lesson store with cross-run retrieval.

    Parameters
    ----------
    evor_root:
        The ``.evor/`` root directory (parent of ``runs/``).
        Cross-run wiki lives at ``evor_root / "wiki"``.
    """

    def __init__(self, evor_root: Path) -> None:
        self._evor_root = Path(evor_root)
        self._wiki_dir = self._evor_root / "wiki"
        self._index_path = self._wiki_dir / "index.jsonl"

    # ── Public API ─────────────────────────────────────────────────────────────

    def add(self, entry: LessonEntry, run_dir: Path) -> str:
        """Persist a lesson to both the per-run wiki and the cross-run index.

        Writes:
          - ``run_dir/wiki/<lesson_id>.md``          (per-run copy)
          - ``evor_root/wiki/<lesson_id>.md``         (cross-run copy)
          - ``evor_root/wiki/index.jsonl``            (appended JSON line)

        Returns the lesson_id for reference chaining.
        """
        self._wiki_dir.mkdir(parents=True, exist_ok=True)

        rendered = self._render_lesson(entry)

        # Cross-run wiki (primary; queried by all subsequent runs)
        (self._wiki_dir / f"{entry.lesson_id}.md").write_text(rendered)
        with open(self._index_path, "a") as fh:
            fh.write(entry.model_dump_json() + "\n")

        # Per-run wiki (convenient when browsing a single run)
        run_wiki = Path(run_dir) / "wiki"
        run_wiki.mkdir(parents=True, exist_ok=True)
        (run_wiki / f"{entry.lesson_id}.md").write_text(rendered)

        return entry.lesson_id

    def query(
        self,
        query: str,
        family: Optional[ApproachFamily] = None,
        confirmed_only: bool = False,
        limit: int = 10,
    ) -> list[LessonEntry]:
        """Keyword scan over the cross-run index.jsonl.

        Matching:
          Each lesson is scored by the number of query keywords that appear in its
          ``observation``, ``actionable_lesson``, ``tags``, or ``root_cause`` text.
          Lessons with zero keyword hits are excluded.

        Filters:
          family       — if provided, only lessons with this approach_family are returned.
          confirmed_only — if True, only lessons where hypothesis_verdict='confirmed'.

        Ranking:
          Results are sorted first by keyword hit count (desc), then by recency
          (created_at desc) to break ties.
        """
        if not self._index_path.exists():
            return []

        keywords = [kw.lower() for kw in query.split() if kw.strip()]

        entries: list[tuple[int, str, LessonEntry]] = []  # (hits, created_at, entry)
        for line in self._index_path.read_text().splitlines():
            line = line.strip()
            if not line:
                continue
            try:
                entry = LessonEntry.model_validate_json(line)
            except Exception:
                continue

            if family is not None and entry.approach_family != family:
                continue
            if confirmed_only and entry.hypothesis_verdict != "confirmed":
                continue

            if not keywords:
                # No keywords → return all (up to limit), ranked by recency
                entries.append((0, entry.created_at, entry))
                continue

            # Build searchable text blob
            tag_text = " ".join(entry.tags)
            blob = " ".join(filter(None, [
                entry.observation,
                entry.actionable_lesson,
                entry.root_cause or "",
                tag_text,
            ])).lower()

            hits = sum(blob.count(kw) for kw in keywords)
            if hits > 0:
                entries.append((hits, entry.created_at, entry))

        # Sort: most keyword hits first, then newest first
        entries.sort(key=lambda x: (x[0], x[1]), reverse=True)
        return [e for _, _, e in entries[:limit]]

    def load_context(self, mission_id: str, limit: int = 5) -> list[LessonEntry]:
        """Return top-N lessons relevant to the mission for Evor context injection.

        Matches lessons whose mission_id equals the given mission_id or whose
        tags/observation mention the mission's task keywords.  Returns the most
        recently created lessons that match.
        """
        if not self._index_path.exists():
            return []

        matched: list[tuple[str, LessonEntry]] = []  # (created_at, entry)
        for line in self._index_path.read_text().splitlines():
            line = line.strip()
            if not line:
                continue
            try:
                entry = LessonEntry.model_validate_json(line)
            except Exception:
                continue

            if entry.mission_id == mission_id:
                matched.append((entry.created_at, entry))

        # Sort newest first
        matched.sort(key=lambda x: x[0], reverse=True)
        return [e for _, e in matched[:limit]]

    def summarize(
        self,
        confirmed_only: bool = False,
        limit: int = 100,
    ) -> dict:
        """Return a summary dict grouping lessons by approach_family and hypothesis_verdict."""
        all_entries = self.query("", confirmed_only=confirmed_only, limit=limit)
        by_family: dict[str, list[str]] = {}
        confirmed = 0
        refuted = 0
        inconclusive = 0
        for entry in all_entries:
            fam = entry.approach_family or "unknown"
            by_family.setdefault(fam, []).append(
                f"[{entry.lesson_id}] {entry.actionable_lesson[:80]} ({entry.hypothesis_verdict})"
            )
            if entry.hypothesis_verdict == "confirmed":
                confirmed += 1
            elif entry.hypothesis_verdict == "refuted":
                refuted += 1
            else:
                inconclusive += 1
        return {
            "confirmed": confirmed,
            "refuted": refuted,
            "inconclusive": inconclusive,
            "by_family": by_family,
        }

    # ── Internal helpers ───────────────────────────────────────────────────────

    def _render_lesson(self, entry: LessonEntry) -> str:
        """Render a LessonEntry as human-readable markdown."""
        lines: list[str] = [
            f"# Lesson: {entry.lesson_id}",
            "",
            f"**Node:** {entry.node_id}  |  **Run:** {entry.run_id}  |  **Mission:** {entry.mission_id}",
            f"**Family:** {entry.approach_family}  |  **Verdict:** {entry.hypothesis_verdict}",
            f"**Created:** {entry.created_at}",
            "",
            "## Observation",
            "",
            entry.observation,
            "",
        ]

        if entry.root_cause:
            lines += ["## Root Cause", "", entry.root_cause, ""]

        lines += [
            "## Actionable Lesson",
            "",
            entry.actionable_lesson,
            "",
        ]

        if entry.telemetry_evidence:
            lines += ["## Telemetry Evidence", "", entry.telemetry_evidence, ""]

        if entry.citations:
            lines += ["## Citations", ""]
            for cit in entry.citations:
                lines.append(f"- {cit}")
            lines.append("")

        if entry.tags:
            lines += [f"**Tags:** {', '.join(entry.tags)}", ""]

        return "\n".join(lines)


# ─────────────────────────────────────────────────────────────────────────────
# CLI entry point — `python -m evor.wiki`
# ─────────────────────────────────────────────────────────────────────────────


def _cli() -> None:
    """CLI for CompoundingWiki.

    Subcommands
    -----------
    query
        Keyword search over the cross-run index.

        python -m evor.wiki query \\
            --query-text "batch size" \\
            [--family arch] [--confirmed-only] [--limit 10] \\
            [--evor-root .evor]

    summarize
        Summarise all lessons grouped by family/verdict.

        python -m evor.wiki summarize \\
            --run-id <id> --run-dir <dir> \\
            [--confirmed-only false] [--evor-root .evor]

    context
        Return top-N lessons for a mission (used by session-start hook).

        python -m evor.wiki context \\
            --mission-id <id> [--limit 5] [--evor-root .evor]
    """
    import argparse
    import json as _json
    import os as _os
    import sys as _sys

    _evor_root_kwargs = dict(
        default=None,
        help="Path to .evor/ root. Defaults to $EVOR_ROOT env var or '.evor'.",
    )

    parser = argparse.ArgumentParser(prog="python -m evor.wiki")
    sub = parser.add_subparsers(dest="cmd", required=True)

    # query subcommand
    q_p = sub.add_parser("query", help="Keyword search over the cross-run index.")
    q_p.add_argument("--query-text", default="", help="Keywords to search for")
    q_p.add_argument("--family", default=None, help="Filter by approach_family")
    q_p.add_argument("--confirmed-only", action="store_true", help="Only confirmed hypotheses")
    q_p.add_argument("--limit", type=int, default=10)
    q_p.add_argument("--evor-root", **_evor_root_kwargs)

    # summarize subcommand
    s_p = sub.add_parser("summarize", help="Summarise lessons by family/verdict.")
    s_p.add_argument("--run-id", default=None)
    s_p.add_argument("--run-dir", default=None)
    s_p.add_argument("--confirmed-only", default="false",
                     help="'true' or 'false' (default false)")
    s_p.add_argument("--limit", type=int, default=100)
    s_p.add_argument("--evor-root", **_evor_root_kwargs)

    # context subcommand
    c_p = sub.add_parser("context", help="Return top-N lessons for a mission.")
    c_p.add_argument("--mission-id", required=True)
    c_p.add_argument("--limit", type=int, default=5)
    c_p.add_argument("--evor-root", **_evor_root_kwargs)

    args = parser.parse_args()

    evor_root = Path(
        args.evor_root
        or _os.environ.get("EVOR_ROOT", "")
        or ".evor"
    )

    wiki = CompoundingWiki(evor_root)

    if args.cmd == "query":
        family = args.family  # already a string or None
        results = wiki.query(
            args.query_text,
            family=family,
            confirmed_only=args.confirmed_only,
            limit=args.limit,
        )
        print(_json.dumps([e.model_dump() for e in results], indent=2))

    elif args.cmd == "summarize":
        confirmed_only_flag = (
            args.confirmed_only.lower() not in ("false", "0", "no")
            if isinstance(args.confirmed_only, str)
            else bool(args.confirmed_only)
        )
        summary = wiki.summarize(confirmed_only=confirmed_only_flag, limit=args.limit)
        print(_json.dumps(summary, indent=2))

    elif args.cmd == "context":
        lessons = wiki.load_context(args.mission_id, limit=args.limit)
        for lesson in lessons:
            print(f"[{lesson.lesson_id}] {lesson.actionable_lesson}")

    _sys.exit(0)


if __name__ == "__main__":
    _cli()
